src/backend/distributed/commands/multi_copy.c

```diff
@@ -986,6 +1165,50 @@ OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections
 
 
 /*
+ * BinaryCopyFormat iterates over columns of the relation given in copyState and
+ * looks for a column whose type is array of user-defined type. If it founds that
+ * means we cannot use binary format for COPY and it return false.
+ */
+static bool
+BinaryCopyFormat(CopyState copyState)
+{
+	TupleDesc tupleDescription = NULL;
+	Form_pg_attribute *attributes = NULL;
+	ListCell *attributeId = NULL;
+	bool binaryCopyFormat = true;
+
+	if (copyState->rel)
+	{
+		tupleDescription = copyState->rel->rd_att;
+	}
+	else
+	{
+		tupleDescription = copyState->queryDesc->tupDesc;
+	}
+
+	attributes = tupleDescription->attrs;
+	foreach(attributeId, copyState->attnumlist)
+	{
+		int attributeNumber = lfirst_int(attributeId);
+		Oid typeId = attributes[attributeNumber - 1]->atttypid;
+		bool builtInType = typeId < FirstNormalObjectId;
+		char typeCategory = '\0';
+		bool typePreferred = false;
+
+		get_type_category_preferred(typeId, &typeCategory, &typePreferred);
+
+		if (!builtInType && typeCategory == TYPCATEGORY_ARRAY)
+		{
+			binaryCopyFormat = false;
+			break;
+		}
```

Comment:
I understand. In case of composite types, oid of the element itself does not included but oids of contained elements are included. Therefore if contained type is not built-in that would cause same problem. I sent a patch to fix this issue also.